src/Type/Php/ArrayMapFunctionReturnTypeExtension.php

```diff
@@ -37,13 +42,27 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
 		$singleArrayArgument = !isset($functionCall->getArgs()[2]);
 		$callableType = $scope->getType($functionCall->getArgs()[0]->value);
 		$callableIsNull = $callableType->isNull()->yes();
+		$callback = null;
 
 		if ($callableType->isCallable()->yes()) {
 			$valueTypes = [new NeverType()];
 			foreach ($callableType->getCallableParametersAcceptors($scope) as $parametersAcceptor) {
 				$valueTypes[] = $parametersAcceptor->getReturnType();
 			}
 			$valueType = TypeCombinator::union(...$valueTypes);
+			$callback = $functionCall->getArgs()[0]->value;
+			if ($callback instanceof String_) {
+				/** @var non-falsy-string $callName */
+				$callName = $callback->value;
+				$callback = new Name($callName);
+			} elseif ($callback instanceof FuncCall && $callback->isFirstClassCallable() &&
+				$callback->getAttribute('phpstan_cache_printer') !== null &&
+				preg_match('/\A(?<name>\\\\?[^()]+)\(...\)\z/', $callback->getAttribute('phpstan_cache_printer'), $m) === 1
```

Comment:
To avoid binding the `ClosureType::$returnType` to a union with the results of other elements, I have replaced it with a direct call to the original function. I don't want to misuse the `phpstan_cache_printer` attribute, but there doesn't seem to be any other suitable source of the original function.